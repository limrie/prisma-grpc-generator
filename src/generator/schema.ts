import { DMMF } from '@prisma/generator-helper'
import { Enum, Field, Method, OneOf, ReflectionObject, Root, Service, Type } from 'protobufjs'
import proto from './protobuf'

const TypeMap = {
    String: 'string',
    Boolean: 'bool',
    Int: 'sint32',
    BigInt: 'sint64',
    Float: 'float',
    DateTime: 'google.protobuf.Timestamp',
    Json: 'google.protobuf.Struct',
    Bytes: 'bytes'
}

type TypeKey = keyof typeof TypeMap

const camelCase = (str : string) => str.substr(0, 1).toLowerCase() + str.substr(1)

const pluralize = (str: string) => str + 's'

const mapEnum = (e : DMMF.DatamodelEnum) : Enum => {
    // ToDo : Keep track of enum field values
    // ToDo : Allow to specify default
    const enumValues = e.values.reduce((agg, value, idx) => ({
        ...agg,
        [value.name]: idx + 1, 
    }), {
        'UNKNOWN': 0
    }) 

    return new Enum(e.name, enumValues)
}

const mapField = (field: DMMF.Field, idx: number) : Field => {
    const rule = field.isList ? 'repeated' : field.isRequired ? 'required' : 'optional'

    let fieldType : string
    switch (field.kind) {
        case 'enum':
        case 'object':
            fieldType = field.type
            break
        case 'scalar':
            if (field.type as TypeKey) {
               fieldType = TypeMap[field.type as TypeKey] 
            } else {
                throw new Error(`Scalar fields of type ${field.type} are not supported by the grpc generator`)
            }
            break
        default:
            throw new Error(`${field.kind} fields are not supported by the grpc generator`)
    }

    return new Field(field.name, idx, fieldType, rule)
}

const mapModel = (model: DMMF.Model) : ReflectionObject[] => {
    const mapped = new Type(model.name)
    const service = new Service(`${model.name}Service`)
    const relationFields = model.fields.map(f => f.relationFromFields).filter(Boolean).flat()

    model.fields.filter(f => !relationFields.includes(f.name)).forEach((field, idx) => {
        mapped.add(mapField(field, idx + 1))
    })

    const camelCased = camelCase(model.name)
    const pluralized = pluralize(model.name)
    const camelCasedPlural = camelCase(pluralized)

    const scalars = new Enum(`${model.name}ScalarField`, model.fields.filter(f => f.kind == 'scalar').map(f => f.name).reduce((agg, f, idx) => ({...agg, [f]: idx + 1}), {UNKNOWN: 0}))

    // CRUD
    // Create One
    const createInput = new Type(`Create${model.name}Input`)
    let idx = 0
    model.fields.filter(f => !f.isGenerated && !relationFields.includes(f.name)).forEach( f => {
        if (f.relationName) {

        } else
            createInput.add(mapField(f, ++idx))
    })

    const creator = new Type(`Create${model.name}Request`)
    creator.add(new Field(camelCased, 1, createInput.name))
    
    const created = new Type(`Create${model.name}Response`)
    created.add(new Field(camelCased, 1, model.name))

    service.add(new Method(`create${model.name}`, 'rpc', creator.name, created.name))
    // Create Many
    const manyCreator = new Type(`CreateMany${pluralized}Request`)
    manyCreator.add(new Field(camelCasedPlural, 1, createInput.name, 'repeated'))
    const manyCreated = new Type(`CreateMany${pluralized}Response`)
    manyCreated.add(new Field(camelCasedPlural, 1, model.name, 'repeated'))

    service.add(new Method(`createMany${pluralized}`, 'rpc', manyCreator.name, manyCreated.name))

    // Find Unique
    const findUniqueRequest = new Type(`FindUnique${model.name}Request`)
    const findUniqueInput = new Type(`${model.name}WhereUniqueInput`)
    const uniqueFields = model.fields.filter(f => f.isUnique || f.isId)
    uniqueFields.forEach((f, idx) => {
        findUniqueInput.add(mapField(f, idx))
    })
    findUniqueInput.add(new OneOf('where', uniqueFields.map(f => f.name)))
    findUniqueRequest.add(findUniqueInput)

    const findUniqueResponse = new Type(`FindUnique${model.name}Response`)
    findUniqueResponse.add(new Field(camelCased, 1, mapped.name))

    service.add(new Method(`findUnique${model.name}`, 'rpc', findUniqueRequest.name, mapped.name))

    // Find Many
    const findManyRequest = new Type(`FindMany${pluralized}Request`)
    const findManyInput = new Type(`${model.name}WhereInput`)
    findManyInput.add(new Field('and', 1, findManyInput.name, 'repeated'))
    findManyInput.add(new Field('or', 2, findManyInput.name, 'repeated'))
    findManyInput.add(new Field('not', 3, findManyInput.name, 'repeated'))
    model.fields.filter(f => f.kind == 'scalar').forEach((field, idx) => {
        findManyInput.add(new Field(field.name, idx + 4, `${field.type}WhereInput`))
    })
    findManyRequest.add(new Field('where', 1, findManyInput.name))  

    const orderByInput = new Type(`${model.name}OrderByInput`)
    const orderables = model.fields
        .filter(field => field.kind == 'scalar' && !relationFields.includes(field.name))
        .map(field => field.name)
    orderables.forEach((f, idx) => {
        orderByInput.add(new Field(f, idx + 1, 'SortOrder'))
    })
    orderByInput.add(new OneOf('field', orderables))
    findManyRequest.add(new Field('orderBy', 2, orderByInput.name, 'repeated'))
    findManyRequest.add(new Field('cursor', 3, findUniqueInput.name))
    findManyRequest.add(new Field('take', 4, 'uint32'))
    findManyRequest.add(new Field('skip', 5, 'int32'))
    findManyRequest.add(new Field('distinct', 6, scalars.name, 'repeated'))

    const findManyResponse = new Type(`FindMany${pluralized}Response`)
    findManyResponse.add(new Field(camelCasedPlural, 1, mapped.name, 'repeated'))

    const findManyStreamResponse = new Type(`FindMany${pluralized}StreamResponse`)
    findManyStreamResponse.add(new Field(camelCased, 1, mapped.name))
    findManyStreamResponse.add(new OneOf('case', [camelCased]))

    service.add(new Method(`findMany${pluralized}`, 'rpc', findManyRequest.name, findManyResponse.name))
    service.add(new Method(`findMany${pluralized}Streamed`, 'rpc', findManyRequest.name, findManyStreamResponse.name, false, true))

    // Delete One
    const deleter = new Type(`Delete${model.name}Request`)
    deleter.add(new Field(camelCased, 1, findUniqueInput.name))
    const deleted = new Type(`Delete${model.name}Response`)
    deleted.add(new Field(camelCased, 1, mapped.name))
    service.add(new Method(`delete${model.name}`, 'rpc', deleter.name, deleted.name))

    const manyDeleter = new Type(`DeleteMany${pluralized}Requests`)
    manyDeleter.add(new Field('where', 1, findManyInput.name))
    const manyDeleted = new Type(`DeleteMany${pluralized}Reponse`)
    manyDeleted.add(new Field('count', 1, 'uint32'))
    service.add(new Method(`deleteMany${pluralized}`, 'rpc', manyDeleter.name, manyDeleted.name))

    return [
        mapped, scalars,
        creator, createInput, created, manyCreator, manyCreated,
        findUniqueRequest, findUniqueInput, findUniqueResponse, 
        findManyRequest, findManyInput, findManyResponse, findManyStreamResponse,
        orderByInput,
        deleter, deleted, manyDeleter, manyDeleted,
        service
    ]
}


export const mapSchema = (doc: DMMF.Document) : Root => {
    const root = new Root()

    // system types
    // sort
    root.add(new Enum('SortOrder', {ASC: 0, DESC: 1}))
    //filter
    doc.datamodel.models.map(model => model.fields
            .filter(f => f.kind == 'scalar')
            .map(f => f.type))
        .flat()
        .reduce((agg : string[], f) => agg.includes(f) ? agg : [...agg, f] , [])
    .forEach(scalar => {
        const scalarWhere = new Type(`${scalar}WhereInput`)
        const protoType = TypeMap[scalar as keyof typeof TypeMap]
        scalarWhere.add(new Field('equals', 1, protoType))
        scalarWhere.add(new Field('in', 2, protoType, 'repeated'))
        root.add(scalarWhere)
    })
    

    doc.datamodel.enums.forEach(e => {
        root.add(mapEnum(e))
    })

    doc.datamodel.models.forEach(model => {
        const operations = doc.mappings.modelOperations.find(mo => mo.model == model.name)
        mapModel(model).forEach(obj => root.add(obj))
    })

    return root
}

export const writeProto = (root: Root, pkg = '') => {
    return proto(root, pkg)
}
