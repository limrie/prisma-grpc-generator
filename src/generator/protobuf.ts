import Protobuf, { Enum, Field, MapField, Method, Namespace, OneOf, ReflectionObject, Root, Service, Type } from 'protobufjs'

const snakecase = (str: string) : string => str.substr(0, 1) 
    + str.substr(1)
    .replace(/([A-Z])(?=[a-z]|$)/g, (_, $1) => '_' + $1.toLowerCase())

const escape = (str: string): string => str
    .replace(/[\\"']/g, "\\$&")
    .replace(/\r/g, "\\r")
    .replace(/\n/g, "\\n")
    .replace(/\u0000/g, "\\0")

const value = (v: any) => {
    switch (typeof v) {
        case "boolean":
            return v ? "true" : "false"
        case "number":
            return v.toString()
        default:
            return "\"" + escape(String(v)) + "\""
    }
}

export const WELL_KNOWN_TYPES = {
    'google.protobuf.Timestamp': 'google/protobuf/timestamp.proto',
    'google.protobuf.Struct': 'google/protobuf/struct.proto'
}

export default async (root: Root, pkg = '', imports = WELL_KNOWN_TYPES) : Promise<string> => {
    const out : string[] = []
    const preamble : string[] = []
    const imported : string[] = []
    let indent = 0

    const importPaths = Object.values(imports).reduce((agg : string[], p) => agg.includes(p) ? agg : [...agg, p], [])

    await root.load(importPaths)

    const push = (line: string) => {
        if (line === "")
            out.push("")
        else {
            let ind = ""
            for (var i = 0; i < indent; ++i)
                ind += "    "
            out.push(ind + line);
        }
    }

    root.resolveAll()

    out.push(`syntax = "proto3";`)
    if(pkg.length) 
      out.push('', `package ${pkg};`)
    buildOptions(root)

    preamble.push(...out)
    out.splice(0, out.length)

    root.nestedArray.forEach(build)

    buildImports()

    return preamble.concat(out).join("\n")


    function build(object: ReflectionObject) {
        if (object instanceof Enum)
            buildEnum(object)
        else if (object instanceof Type)
            buildType(object)
        else if (object instanceof Field)
            buildField(object)
        else if (object instanceof OneOf)
            buildOneOf(object)
        else if (object instanceof Service)
            buildService(object)
        else if (object instanceof Method)
            buildMethod(object)
        else if (object instanceof Namespace)
            buildNamespace(object)
        else 
          throw new Error("Unexpected reflection object of type " + (typeof object))
    }

    function buildEnum(obj: Enum) {
        push("")
        push(`enum ${obj.name} {`)
        ++indent
        const built = buildOptions(obj)
        if(built)
            push("")
        Object.keys(obj.values).forEach(name => {
            const val = obj.values[name]
            push(`${name} = ${val};`)
        })
        --indent
        push("")
    }

    function buildType(obj: Type) {
        const oneOfFields = obj.oneofsArray.map(ofs => ofs.fieldsArray.map(f => f.name)).flat()

        push("")
        push(`message ${obj.name} {`)
        ++indent
        const built = buildOptions(obj)
        if (built) 
            push('')
        obj.oneofsArray.forEach(build)
        obj.fieldsArray.filter(field => !oneOfFields.includes(field.name)).forEach(build)
        obj.nestedArray.forEach(build)
        buildRanges('reserved', obj.reserved)
        --indent
        push("}")
    }

    function buildField(obj: Field) {
        const sb : string[] = []
        if(obj instanceof MapField) 
            sb.push(`map<${obj.keyType}, ${obj.type}>`)
        else if(obj.repeated) 
            sb.push(`repeated ${obj.type}`)
        else
            sb.push(obj.type)
        sb.push(`${snakecase(obj.name)} = ${obj.id}`)
        const opts = buildFieldOptions(obj)
        if(opts)
          sb.push(opts)
        
        push(sb.join(" ") + ';')

        if(imports.hasOwnProperty(obj.type) && !imported.includes(imports[obj.type as keyof typeof imports])) 
            imported.push(imports[obj.type as keyof typeof imports])
    }

    function buildFieldOptions(obj: Field) : string | void | null {
        const opts = obj.options
        if(opts) {
            const sb: string[] = []
            Object.keys(opts).forEach(key => {
                let val = opts[key]
                const typeKey = obj.resolvedType instanceof Enum ? 'int32' : obj.type
                const wireType = Protobuf.types.packed[typeKey as (keyof typeof Protobuf.types.packed)]
                switch(key) {
                    case 'packed':
                        val = Boolean(val)
                        if(wireType === undefined || val)
                            return
                        break
                    case 'default':
                        return
                    default:
                        val = value(val)
                    break
                }
                sb.push(`${key} = ${val}`)
            })

            return sb.length ? `[${sb.join(', ')}]` : null
        }

        return
    }

    function buildOneOf(obj: OneOf) {
        push(`oneof ${snakecase(obj.name)} {`)
        ++indent
        obj.oneof.forEach(fieldname => {
            const field = obj.parent?.get(fieldname)
            if(!(field instanceof Field)) {
                throw new Error(`Unable to find field ${fieldname} for oneof ${obj.name}`)
            } else {
                const opts = buildFieldOptions(field)
                push(`${field.type} ${snakecase(field.name)} = ${field.id}${opts ? ` ${opts}` : ''};`)
            }
        })
        --indent
        push('}')
    }

    function buildService(obj: Service) {
        push(`service ${obj.name} {`)
        ++indent
        obj.methodsArray.forEach(build)
        obj.nestedArray.forEach(build)
        --indent
        push('}')
    }

    function buildMethod(obj: Method) {
        push(`${obj.type} ${obj.name} (${obj.requestStream ? 'stream ' : ''}${obj.requestType}) returns (${obj.responseStream ? 'stream ' : ''}${obj.responseType}) {}`)
    }

    function buildNamespace(obj: Namespace) {
        // skip - this should mean something has come from an import
    }

    function buildOptions(obj: ReflectionObject) {
        const options = obj.options
        if (options && options.length) {
            push("")
            Object.keys(options).forEach(key => {
                push(`option ${key} = ${JSON.stringify(options[key])}`)
            })
            return true
        }
        return false
    }

    function buildRanges(keyword : string, ranges : (string | number[])[]) {
        if (ranges && ranges.length) {
            let parts : string[] = []
            ranges.forEach(function (range) {
                if (typeof range === "string")
                    parts.push(`"${escape(range)}"`)
                else if (range[0] === range[1])
                    parts.push(`${range[0]}`);
                else
                    parts.push(`${range[0]} to ${range[1] === 0x1FFFFFFF ? 'max' : range[1]}`)
            })
            push("")
            push(`${keyword} ${parts.join(', ')};`)
        }
    }

    function buildImports() {
        if (imported.length)
            preamble.push('')

        imported.forEach(path => {
            preamble.push(`import "${path}";`)
        })
    }
}