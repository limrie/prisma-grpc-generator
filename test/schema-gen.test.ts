import { mapSchema, writeProto } from '@/generator/schema'
import { getDMMF } from '@prisma/sdk'
import { join } from 'path'

test('generate grpc schema', async () => {
    const dmmf = await getDMMF({ datamodelPath: join(__dirname, './test.prisma') })
    const schema = mapSchema(dmmf)
    const proto = await writeProto(schema)
    expect(proto).toMatchSnapshot()
})