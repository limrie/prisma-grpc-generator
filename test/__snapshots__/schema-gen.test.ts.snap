// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate grpc schema 1`] = `
"syntax = \\"proto3\\";

import \\"google/protobuf/timestamp.proto\\";

enum SortOrder {
    ASC = 0;
    DESC = 1;


message StringWhereInput {
    string equals = 1;
    repeated string in = 2;
}

message DateTimeWhereInput {
    google.protobuf.Timestamp equals = 1;
    repeated google.protobuf.Timestamp in = 2;
}

message IntWhereInput {
    sint32 equals = 1;
    repeated sint32 in = 2;
}

message User {
    string id = 1;
    google.protobuf.Timestamp created_at = 2;
    string first_name = 3;
    string last_name = 4;
    string email = 5;
    repeated Post posts = 6;
}

enum UserScalarField {
    UNKNOWN = 0;
    id = 1;
    createdAt = 2;
    firstName = 3;
    lastName = 4;
    email = 5;


message CreateUserRequest {
    CreateUserInput user = 1;
}

message CreateUserInput {
    string id = 1;
    google.protobuf.Timestamp created_at = 2;
    string first_name = 3;
    string last_name = 4;
    string email = 5;
}

message CreateUserResponse {
    User user = 1;
}

message CreateManyUsersRequest {
    repeated CreateUserInput users = 1;
}

message CreateManyUsersResponse {
    repeated User users = 1;
}

message FindUniqueUserRequest {
}

message UserWhereUniqueInput {
    oneof where {
        string id = 0;
        string email = 1;
    }
}

message FindUniqueUserResponse {
    User user = 1;
}

message FindManyUsersRequest {
    UserWhereInput where = 1;
    repeated UserOrderByInput order_by = 2;
    UserWhereUniqueInput cursor = 3;
    uint32 take = 4;
    int32 skip = 5;
    repeated UserScalarField distinct = 6;
}

message UserWhereInput {
    repeated UserWhereInput and = 1;
    repeated UserWhereInput or = 2;
    repeated UserWhereInput not = 3;
    StringWhereInput id = 4;
    DateTimeWhereInput created_at = 5;
    StringWhereInput first_name = 6;
    StringWhereInput last_name = 7;
    StringWhereInput email = 8;
}

message FindManyUsersResponse {
    repeated User users = 1;
}

message FindManyUsersStreamResponse {
    oneof case {
        User user = 1;
    }
}

message UserOrderByInput {
    oneof field {
        SortOrder id = 1;
        SortOrder created_at = 2;
        SortOrder first_name = 3;
        SortOrder last_name = 4;
        SortOrder email = 5;
    }
}

message DeleteUserRequest {
    UserWhereUniqueInput user = 1;
}

message DeleteUserResponse {
    User user = 1;
}

message DeleteManyUsersRequests {
    UserWhereInput where = 1;
}

message DeleteManyUsersReponse {
    uint32 count = 1;
}
service UserService {
    rpc createUser (CreateUserRequest) returns (CreateUserResponse) {}
    rpc createManyUsers (CreateManyUsersRequest) returns (CreateManyUsersResponse) {}
    rpc findUniqueUser (FindUniqueUserRequest) returns (User) {}
    rpc findManyUsers (FindManyUsersRequest) returns (FindManyUsersResponse) {}
    rpc findManyUsersStreamed (FindManyUsersRequest) returns (stream FindManyUsersStreamResponse) {}
    rpc deleteUser (DeleteUserRequest) returns (DeleteUserResponse) {}
    rpc deleteManyUsers (DeleteManyUsersRequests) returns (DeleteManyUsersReponse) {}
}

message Post {
    string id = 1;
    google.protobuf.Timestamp created_at = 2;
    string title = 3;
    string body = 4;
    sint32 views = 5;
    User author = 6;
}

enum PostScalarField {
    UNKNOWN = 0;
    id = 1;
    createdAt = 2;
    title = 3;
    body = 4;
    views = 5;
    authorId = 6;


message CreatePostRequest {
    CreatePostInput post = 1;
}

message CreatePostInput {
    string id = 1;
    google.protobuf.Timestamp created_at = 2;
    string title = 3;
    string body = 4;
    sint32 views = 5;
}

message CreatePostResponse {
    Post post = 1;
}

message CreateManyPostsRequest {
    repeated CreatePostInput posts = 1;
}

message CreateManyPostsResponse {
    repeated Post posts = 1;
}

message FindUniquePostRequest {
}

message PostWhereUniqueInput {
    oneof where {
        string id = 0;
    }
}

message FindUniquePostResponse {
    Post post = 1;
}

message FindManyPostsRequest {
    PostWhereInput where = 1;
    repeated PostOrderByInput order_by = 2;
    PostWhereUniqueInput cursor = 3;
    uint32 take = 4;
    int32 skip = 5;
    repeated PostScalarField distinct = 6;
}

message PostWhereInput {
    repeated PostWhereInput and = 1;
    repeated PostWhereInput or = 2;
    repeated PostWhereInput not = 3;
    StringWhereInput id = 4;
    DateTimeWhereInput created_at = 5;
    StringWhereInput title = 6;
    StringWhereInput body = 7;
    IntWhereInput views = 8;
    StringWhereInput author_id = 9;
}

message FindManyPostsResponse {
    repeated Post posts = 1;
}

message FindManyPostsStreamResponse {
    oneof case {
        Post post = 1;
    }
}

message PostOrderByInput {
    oneof field {
        SortOrder id = 1;
        SortOrder created_at = 2;
        SortOrder title = 3;
        SortOrder body = 4;
        SortOrder views = 5;
    }
}

message DeletePostRequest {
    PostWhereUniqueInput post = 1;
}

message DeletePostResponse {
    Post post = 1;
}

message DeleteManyPostsRequests {
    PostWhereInput where = 1;
}

message DeleteManyPostsReponse {
    uint32 count = 1;
}
service PostService {
    rpc createPost (CreatePostRequest) returns (CreatePostResponse) {}
    rpc createManyPosts (CreateManyPostsRequest) returns (CreateManyPostsResponse) {}
    rpc findUniquePost (FindUniquePostRequest) returns (Post) {}
    rpc findManyPosts (FindManyPostsRequest) returns (FindManyPostsResponse) {}
    rpc findManyPostsStreamed (FindManyPostsRequest) returns (stream FindManyPostsStreamResponse) {}
    rpc deletePost (DeletePostRequest) returns (DeletePostResponse) {}
    rpc deleteManyPosts (DeleteManyPostsRequests) returns (DeleteManyPostsReponse) {}
}"
`;
